raiden/raiden_service.py

```diff
@@ -212,108 +287,268 @@ def register_secret(self, secret):
-        """ Register the registry and intialize all the related tokens and
-        channels.
+        if channel not in channels_registered:
+            channels_registered.append(channel)
+
+    def handle_secret(  # pylint: disable=too-many-arguments
+            self,
+            identifier,
+            token_address,
+            secret,
+            partner_secret_message,
+            hashlock):
+        """ Unlock/Witdraws locks, register the secret, and send Secret
+        messages as necessary.
+
+        This function will:
+            - Unlock the locks created by this node and send a Secret message to
+            the corresponding partner so that she can withdraw the token.
+            - Withdraw the lock from sender.
+            - Register the secret for the locks received and reveal the secret
+            to the senders
+
+        Note:
+            The channel needs to be registered with
+            `raiden.register_channel_for_hashlock`.
         """
-        translator = ContractTranslator(REGISTRY_ABI)
+        # handling the secret needs to:
+        # - unlock the token for all `forward_channel` (the current one
+        #   and the ones that failed with a refund)
+        # - send a message to each of the forward nodes allowing them
+        #   to withdraw the token
+        # - register the secret for the `originating_channel` so that a
+        #   proof can be made, if necessary
+        # - reveal the secret to the `sender` node (otherwise we
+        #   cannot withdraw the token)
+        channels_list = self.hashlock_channel[token_address][hashlock]
+        channels_to_remove = list()
+
+        # Dont use the partner_secret_message.token since it might not match
+        # the current token manager
+        our_secret_message = Secret(
+            identifier,
+            secret,
+            token_address,
+        )
+        self.sign(our_secret_message)
+
+        revealsecret_message = RevealSecret(secret)
+        self.sign(revealsecret_message)
+
+        for channel in channels_list:
+            # unlock a sent lock
+            if channel.partner_state.balance_proof.is_unclaimed(hashlock):
+                channel.release_lock(secret)
+                self.send_async(
+                    channel.partner_state.address,
+                    our_secret_message,
+                )
+                channels_to_remove.append(channel)
+
+            # withdraw a pending lock
+            if channel.our_state.balance_proof.is_unclaimed(hashlock):
+                if partner_secret_message:
+                    valid_sender = partner_secret_message.sender == channel.partner_state.address
+                    valid_token = partner_secret_message.token == channel.token_address
+
+                    if valid_sender and valid_token:
+                        channel.withdraw_lock(secret)
+                        channels_to_remove.append(channel)
+                    else:
```

Comment:
Note: This method is a legacy from the greenlet based tasks, I plan to remove it in the future, but for now it's required for the token swap.

As for the `else`, the secret is valid since the hashlock matches the list of channels in `hashlock_channel`, so we are just revealing it to the other nodes.